### PR TITLE
Fix Async.Parallel stack overflow on cancellation of ~2000 uncompleted

### PR DESCRIPTION
**Pull request by dsyme**
Fixes https://github.com/dotnet/fsharp/issues/13165

### CODE REVIEW

**Review by vzarytovskii — APPROVED** [review]

**Comment by dsyme** [comment]
Transient failure: `Move-Item : Access to the path 'C:\Users\cloudtest\.netcoreeng\native\bin\perl\5.32.1.1.tmp' is denied.`

File: tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Control/AsyncModule.fs [hunk 1]
```diff
@@ -764,3 +764,15 @@ type AsyncModule() =
...
+    [<Fact>]
+    member _.``Async.Parallel blows stack when cancelling many`` () =
+        let gen (i : int) = async {
+            if i <> 0 then do! Async.Sleep i
+            else return failwith (string i) }
+        let count = 1800
+        let comps = Seq.init count gen
+        let result = Async.Parallel(comps, 16) |> Async.Catch |> Async.RunSynchronously
+        match result with
+        | Choice2Of2 e -> ()
```
Comment (bartelink) [inline tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Control/AsyncModule.fs:777]: perhaps this assertion could be strengthened (maybe by throwing something more specific at L772?)

**Comment by vzarytovskii** [comment]
One leg is failing with:
> Move-Item : Access to the path 'C:\Users\cloudtest\.netcoreeng\native\bin\perl\5.32.1.1.tmp' is denied.

AFAIU dnceng are working on it, @brettfo is that correct?

File: tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Control/AsyncModule.fs [hunk 2]
```diff
@@ -764,3 +764,15 @@ type AsyncModule() =
                 lock gate <| fun () -> printfn "Unhandled exception: %s" exn.Message
                 lock gate <| fun () -> printfn "Semaphore count available: %i" semaphore.CurrentCount
             Assert.AreEqual(acquiredCount, releaseCount)
+
+    [<Fact>]
+    member _.``Async.Parallel blows stack when cancelling many`` () =
+        let gen (i : int) = async {
+            if i <> 0 then do! Async.Sleep i
+            else return failwith (string i) }
+        let count = 1800
```
[the remark this answers is not in the document]
Comment (dsyme) [reply tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Control/AsyncModule.fs]: Doubled!

File: tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Control/AsyncModule.fs (hunk 1 above)
Comment (dsyme) [reply tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Control/AsyncModule.fs]: Updated (editing on my phone, may have made a mistake!)

File: tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Control/AsyncModule.fs [hunk 3]
```diff
@@ -764,3 +764,15 @@ type AsyncModule() =
                 lock gate <| fun () -> printfn "Unhandled exception: %s" exn.Message
                 lock gate <| fun () -> printfn "Semaphore count available: %i" semaphore.CurrentCount
             Assert.AreEqual(acquiredCount, releaseCount)
+
+    [<Fact>]
+    member _.``Async.Parallel blows stack when cancelling many`` () =
```
Comment (bartelink) [inline tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Control/AsyncModule.fs:769]:
```suggestion
    // versions 4.7 - 6.0.4 don't tail recurse and hence can blow the stack given ~1600 excess computations at point of cancellation
    member _.``Async.Parallel handles early cancelation of large computation streams safely`` () =
```

**Pull request merged**